### PR TITLE
*: Add initial remote-write-receive component

### PR DESCRIPTION
**Pull request by brancz**
## Changes

This adds the very first pieces of the [proposal to add receiving remote-write](https://github.com/improbable-eng/thanos/blob/master/docs/proposals/approved/201812_thanos-remote-receive.md) support to Thanos. The features are just having a single tsdb that remote write requests are written into and that the component joins a mesh network to enable querying. All other features are still outstanding but can be worked on in parallel once this lands.

## Verification

As this is very small and primarily hooks up existing components, the extend of my verification is that I added this new component to the quickstart script and verified that indeed data replicated to the endpoint receiving remote-write calls can be queried via the Thanos querier. Happy to add this to the e2e test suite though.

@bwplotka @domgreen 

cc @metalmatze @mxinden @squat @s-urbaniak @ant31

### CODE REVIEW

**Comment by bwplotka** [comment]
Finally <3

First of all, as you can see our CI does not like NOT checked errors (: and both:
```
fmt.Fprintf	fmt.Fprintf(w, "Service Unavailable")
(*net/http.Server).Shutdown	httpSrv.Shutdown(ctx)
```
looks like it would be nice at least log them.

**Comment by brancz** [comment]
Looks like that made CI happy :slightly_smiling_face: .

**Comment by brancz** [comment]
I think having a simple e2e test added would be good. I'll look into adding one.

File: cmd/thanos/remote-write-receive.go [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/storage/tsdb"
+	"google.golang.org/grpc"
+	kingpin "gopkg.in/alecthomas/kingpin.v2"
+)
+
+func registerRemoteWriteReceive(m map[string]setupFunc, app *kingpin.Application, name string) {
+	cmd := app.Command(name, "accept Prometheus remote write API requests (EXPERIMENTAL, this may change drastically without notice)")
```
Comment (bwplotka) [inline cmd/thanos/remote-write-receive.go]:
```suggestion
	cmd := app.Command(name, "Accept Prometheus remote write API requests (EXPERIMENTAL, this may change drastically without notice)")
```

File: cmd/thanos/remote-write-receive.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+				reloadReady.Close()
+
+				webHandler.Ready()
+				level.Info(logger).Log("msg", "Server is ready to receive web requests.")
+				<-cancel
+				return nil
+			},
+			func(err error) {
+				level.Debug(logger).Log("msg", "initial load group errored", "err", err)
```
Comment (bwplotka) [inline cmd/thanos/remote-write-receive.go]: hmm, I don't get this log messages here and on every group. This code is invoked even though this group did not errof (but e.g other group errored). I think we can remove those debug logs. We will print error for problematic group via `g.Run` anyway.

File: cmd/thanos/remote-write-receive.go [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+
+				startTimeMargin := int64(2 * time.Duration(tsdbCfg.MinBlockDuration).Seconds() * 1000)
+				localStorage.Set(db, startTimeMargin)
+				close(dbOpen)
+				<-cancel
+				return nil
+			},
+			func(err error) {
+				level.Debug(logger).Log("msg", "tsdb group errored", "err", err)
+				if err := localStorage.Close(); err != nil {
```
Comment (bwplotka) [inline cmd/thanos/remote-write-receive.go]: just checking .. localStorage.Close closes `db` as well?

File: cmd/thanos/remote-write-receive.go [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+					level.Error(logger).Log("msg", "Error stopping storage", "err", err)
```
Comment (bwplotka) [inline cmd/thanos/remote-write-receive.go]: all log lines should start lowercase

File: cmd/thanos/remote-write-receive.go [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+				}
+				close(cancel)
+			},
+		)
+	}
+
+	level.Debug(logger).Log("msg", "setting up metric http listen-group")
+	if err := metricHTTPListenGroup(g, logger, reg, httpMetricsBindAddr); err != nil {
+		level.Debug(logger).Log("msg", "metric listener errored", "err", err)
```
Comment (bwplotka) [inline cmd/thanos/remote-write-receive.go]: Replicated error handling. (rule "Handle error just once") (: 

I think error wrap is the way to go (and removal of this log line).

File: cmd/thanos/remote-write-receive.go [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+	}
+
+	level.Debug(logger).Log("msg", "setting up tsdb")
+	{
+		// TSDB.
+		cancel := make(chan struct{})
+		g.Add(
+			func() error {
+				level.Info(logger).Log("msg", "Starting TSDB ...")
+				db, err := tsdb.Open(
```
Comment (bwplotka) [inline cmd/thanos/remote-write-receive.go]: hm.. can we open DB outside of run group? And avoid blocking on `dbOpen` entirely? (:

File: pkg/receive/handler.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+	Receiver       *ReceiveWriter
+	Context        context.Context
+	ListenAddress  string
+	MaxConnections int
+	ReadTimeout    time.Duration
+	Registry       prometheus.Registerer
+	ReadyStorage   *tsdb.ReadyStorage
+}
+
+// Handler serves various HTTP endpoints of the Prometheus server
```
Comment (bwplotka) [inline pkg/receive/handler.go]: Missing trailing period. Also `various` is kind of wide word..  I think we expose only `receive` right? (:

File: pkg/receive/handler.go [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+	errlog := stdlog.New(log.NewStdlibAdapter(level.Error(h.logger)), "", 0)
+
+	httpSrv := &http.Server{
+		Handler:     nethttp.Middleware(opentracing.GlobalTracer(), mux, operationName),
+		ErrorLog:    errlog,
+		ReadTimeout: h.options.ReadTimeout,
+	}
+
+	errCh := make(chan error)
```
Comment (bwplotka) [inline pkg/receive/handler.go]: I am pretty sure we leak go routine here, as we send twice but read only once.

Worth to use `run.Group` here for easier code?

File: pkg/receive/handler.go [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+	listener = netutil.LimitListener(listener, h.options.MaxConnections)
+
+	// Monitor incoming connections with conntrack.
+	listener = conntrack.NewListener(listener,
+		conntrack.TrackWithName("http"),
+		conntrack.TrackWithTracing())
+
+	var (
+		m     = cmux.New(listener)
+		httpl = m.Match(cmux.HTTP1Fast())
```
Comment (bwplotka) [inline pkg/receive/handler.go]: Can you comment what this is for? It might be not straightforward.

File: pkg/receive/handler.go [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+		return
+	}
+
+	var wreq prompb.WriteRequest
+	if err := proto.Unmarshal(reqBuf, &wreq); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	err = h.receiver.Receive(&wreq)
```
Comment (bwplotka) [inline pkg/receive/handler.go]: we can do `if err := ... ; err != nil` as well

File: pkg/receive/receivewriter.go [hunk 11]
```diff
@@ -0,0 +1,57 @@
+package receive
+
+import (
+	"github.com/go-kit/kit/log"
```
Comment (bwplotka) [inline pkg/receive/receivewriter.go]: `make format` needed (:

File: scripts/prometheus.yml [hunk 12]
```diff
@@ -0,0 +1,10 @@
+# When the Thanos remote-write-receive component is started,
```
Comment (bwplotka) [inline scripts/prometheus.yml]: Hm.. can we add this to `remote-write-receive` command description instead?  Might be easier to discover

File: scripts/quickstart.sh [hunk 13]
```diff
@@ -136,4 +142,22 @@ do
     --cluster.peers             127.0.0.1:19391 &
 done
 
+sleep 0.5
+
+if [ -n "${REMOTE_WRITE_ENABLED}" ]
```
Comment (bwplotka) [inline scripts/quickstart.sh]: Wow I did not know that this script still works (:

**Comment by bwplotka** [comment]
And yea, some tests would be nice. (:

File: cmd/thanos/remote-write-receive.go [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+	remoteWriteAddress string,
+	dataDir string,
+	peer cluster.Peer,
+	component string,
+) error {
+	level.Info(logger).Log("msg", "setting up receiver")
+
+	var metadata = &promMetadata{
+		// Start out with the full time range. The shipper will constrain it later.
+		// TODO(fabxc): minimum timestamp is never adjusted if shipping is disabled.
```
Comment (metalmatze) [inline cmd/thanos/remote-write-receive.go]: Was this code copied from somewhere? Slightly confusing to see `TODO(fabxc)` in here, but not a big deal I guess.

Comment (bwplotka) [reply cmd/thanos/remote-write-receive.go]: haha yea was surprised as well at  the beginning (: It is copied indeed.

File: cmd/thanos/remote-write-receive.go (hunk 2 above)
Comment (brancz) [reply cmd/thanos/remote-write-receive.go]: Sorry, this is a leftover from some debugging, I'll remove it.

File: pkg/receive/handler.go [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+			Buckets: prometheus.ExponentialBuckets(100, 10, 8),
+		},
+		[]string{"handler"},
+	)
+)
+
+// Options for the web Handler.
+type Options struct {
+	Receiver       *ReceiveWriter
+	Context        context.Context
```
Comment (metalmatze) [inline pkg/receive/handler.go]: This can be removed as it is not used.

File: pkg/receive/handler.go [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+	MaxConnections int
+	ReadTimeout    time.Duration
+	Registry       prometheus.Registerer
+	ReadyStorage   *tsdb.ReadyStorage
+}
+
+// Handler serves various HTTP endpoints of the Prometheus server
+type Handler struct {
+	readyStorage *tsdb.ReadyStorage
+	context      context.Context
```
Comment (metalmatze) [inline pkg/receive/handler.go]: This can be removed as it is not used.

File: pkg/receive/handler.go (hunk 8 above)
Comment (metalmatze) [reply pkg/receive/handler.go]: Yes, this should be a `run.Group`.

File: pkg/receive/handler.go [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+		return e
+	case <-ctx.Done():
+		return httpSrv.Shutdown(ctx)
+	}
+}
+
+func (h *Handler) receive(w http.ResponseWriter, req *http.Request) {
+	defer func() {
+		if r := recover(); r != nil {
+			fmt.Println("panic recovered:", r, string(debug.Stack()))
```
Comment (metalmatze) [inline pkg/receive/handler.go]: Maybe we want to have a counter for how many panics were recovered?

File: pkg/receive/receivewriter.go [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+)
+
+// Appendable returns an Appender.
+type Appendable interface {
+	Appender() (storage.Appender, error)
+}
+
+type ReceiveWriter struct {
```
Comment (metalmatze) [inline pkg/receive/receivewriter.go]: This is spelled `receive.ReceiveWriter` and should be renamed to `receive.Writer` to be more idiomatic.

File: cmd/thanos/remote-write-receive.go (hunk 3 above)
Comment (brancz) [reply cmd/thanos/remote-write-receive.go]: it does

File: cmd/thanos/remote-write-receive.go (hunk 6 above)
Comment (brancz) [reply cmd/thanos/remote-write-receive.go]: The point here is that we respond to http requests as early as possible, but with `StatusServiceUnavailable` until tsdb is loaded.

**Comment by brancz** [comment]
@bwplotka @metalmatze I believe I addressed everything or answered the comments and added the receiver to the e2e test setup (the test is that a prometheus is successfully pushing an `up` metric of a "down" node-exporter).

File: cmd/thanos/main.go [hunk 19]
```diff
@@ -78,6 +78,7 @@ func main() {
 	registerCompact(cmds, app, "compact")
 	registerBucket(cmds, app, "bucket")
 	registerDownsample(cmds, app, "downsample")
+	registerRemoteWriteReceive(cmds, app, "remote-write-receive")
```
Comment (domgreen) [inline cmd/thanos/main.go]: nit - would shorten this to just `receive`/`remote` /`receiver`

File: cmd/thanos/remote-write-receive.go [hunk 20]
```diff
@@ -0,0 +1,270 @@
...
+				}
+			}
+
+			<-ctx.Done()
+			return nil
+		}, func(err error) {
+			cancel()
+			peer.Close(2 * time.Second)
+		})
+	}
```
Comment (domgreen) [inline cmd/thanos/remote-write-receive.go]: can we remove this whole block? Will receive need to peer via gossip? If we are removing gossip in the near future I think we should not add it to any new components and instead help people migrate to the new approach.

File: cmd/thanos/remote-write-receive.go [hunk 21]
```diff
@@ -0,0 +1,270 @@
...
+
+	level.Debug(logger).Log("msg", "setting up metric http listen-group")
+	if err := metricHTTPListenGroup(g, logger, reg, httpMetricsBindAddr); err != nil {
+		return err
+	}
+
+	level.Debug(logger).Log("msg", "setting up grpc server")
+	{
+		var (
+			logger = log.With(logger, "component", "receiver")
```
Comment (domgreen) [inline cmd/thanos/remote-write-receive.go]: `receiver` > `receive`? or the verbose name? Let's keep it consistant.

File: pkg/component/component.go [hunk 22]
```diff
@@ -71,17 +71,20 @@ func FromProto(storeType storepb.StoreType) StoreAPI {
...
-	Sidecar    = sourceStoreAPI{component: component{name: "sidecar"}}
-	Store      = sourceStoreAPI{component: component{name: "store"}}
+	Bucket      = source{component: component{name: "bucket"}}
+	Compact     = source{component: component{name: "compact"}}
+	Downsample  = source{component: component{name: "downsample"}}
+	Query       = sourceStoreAPI{component: component{name: "query"}}
+	Rule        = sourceStoreAPI{component: component{name: "rule"}}
+	Sidecar     = sourceStoreAPI{component: component{name: "sidecar"}}
+	Store       = sourceStoreAPI{component: component{name: "store"}}
+	RemoteWrite = sourceStoreAPI{component: component{name: "remote_write"}}
```
Comment (domgreen) [inline pkg/component/component.go]: consistency

File: pkg/receive/handler.go [hunk 23]
```diff
@@ -0,0 +1,197 @@
...
+func (h *Handler) receive(w http.ResponseWriter, req *http.Request) {
+	compressed, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	reqBuf, err := snappy.Decode(nil, compressed)
+	if err != nil {
+		fmt.Println("snappy decode error")
```
Comment (domgreen) [inline pkg/receive/handler.go]: `fmt` > `logger`?

File: scripts/quickstart.sh [hunk 24]
```diff
@@ -136,4 +142,35 @@ do
...
+then
+  ./thanos remote-write-receive \
```
Comment (domgreen) [inline scripts/quickstart.sh]: :+1: this is great taht were adding it here as well.

File: cmd/thanos/remote-write-receive.go [hunk 25]
```diff
@@ -0,0 +1,270 @@
...
+	grpcBindAddr string,
+	cert string,
+	key string,
+	clientCA string,
+	httpMetricsBindAddr string,
+	remoteWriteAddress string,
+	dataDir string,
+	peer cluster.Peer,
+) error {
+	level.Info(logger).Log("msg", "setting up receiver")
```
Comment (domgreen) [inline cmd/thanos/remote-write-receive.go]: Is it worth adding a warn log here saying that it is currently experimental?

Comment (brancz) [reply cmd/thanos/remote-write-receive.go]: very good idea :+1:

File: scripts/quickstart.sh (hunk 24 above)
Comment (brancz) [reply scripts/quickstart.sh]: either we leave this or remove gossip :slightly_smiling_face:

File: pkg/receive/handler.go [hunk 26]
```diff
@@ -0,0 +1,203 @@
...
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/common/route"
+	promtsdb "github.com/prometheus/prometheus/storage/tsdb"
+	"github.com/prometheus/tsdb"
+)
+
+var (
+	requestDuration = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "prometheus_http_request_duration_seconds",
```
Comment (bwplotka) [inline pkg/receive/handler.go]: We might want `thanos` prefix for those metrics? Or do we want to keep it similar to get advantage of common dashboards?

File: pkg/receive/handler.go [hunk 27]
```diff
@@ -0,0 +1,203 @@
...
+	ready := atomic.LoadUint32(&h.ready)
+	return ready > 0
+}
+
+// Checks if server is ready, calls f if it is, returns 503 if it is not.
+func (h *Handler) testReady(f http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if h.isReady() {
+			f(w, r)
+		} else {
```
Comment (bwplotka) [inline pkg/receive/handler.go]: just return here, no need for `else`

File: pkg/receive/handler.go (hunk 8 above)
Comment (bwplotka) [reply pkg/receive/handler.go]: still valid?

File: scripts/quickstart.sh (hunk 24 above)
Comment (domgreen) [reply scripts/quickstart.sh]: @bwplotka what do you think on this ... personally, I think we should phase out gossip so (in a different PR) change this file to work via the new approach. 
So for this PR I would remove `receive` from this and once the above work has been done add it back in.

File: pkg/receive/handler.go (hunk 26 above)
Comment (brancz) [reply pkg/receive/handler.go]: Sorry that was a copy paste mistake. Yes lets make it `thanos_http_...`.

File: scripts/quickstart.sh (hunk 24 above)
Comment (brancz) [reply scripts/quickstart.sh]: As we have the e2e tests, that works for me.

Comment (brancz) [reply scripts/quickstart.sh]: As there is no git-history of this, I'm going to comment this out for now, but leave it. Does that sound like an ok compromise?

Comment (domgreen) [reply scripts/quickstart.sh]: I'm happy with that ... can you make sure a comment about why its commented is included :smile:

Comment (brancz) [reply scripts/quickstart.sh]: Sounds good to me

File: cmd/thanos/receive.go [hunk 28]
```diff
@@ -0,0 +1,271 @@
...
+	grpcBindAddr, httpMetricsBindAddr, cert, key, clientCA, newPeerFn := regCommonServerFlags(cmd)
+
+	remoteWriteAddress := cmd.Flag("remote-write.address", "Address to listen on for remote write requests.").
+		Default("0.0.0.0:19291").String()
+
+	dataDir := cmd.Flag("tsdb.path", "Data directory of TSDB.").
+		Default("./data").String()
+
+	m[name] = func(g *run.Group, logger log.Logger, reg *prometheus.Registry, tracer opentracing.Tracer, _ bool) error {
+		peer, err := newPeerFn(logger, reg, false, "", false)
```
Comment (bwplotka) [inline cmd/thanos/receive.go]: You can actualy remove gossip stuff if you want, up to you - making sure deprecated stuff is in new component is quite not necessary. (:  Not a blocker

File: cmd/thanos/receive.go [hunk 29]
```diff
@@ -0,0 +1,271 @@
...
+	grpcBindAddr string,
+	cert string,
+	key string,
+	clientCA string,
+	httpMetricsBindAddr string,
+	remoteWriteAddress string,
+	dataDir string,
+	peer cluster.Peer,
+) error {
+	level.Info(logger).Log("msg", "setting up receive")
```
Comment (bwplotka) [inline cmd/thanos/receive.go]: Let's kill this message, below as `setting up receive; this component is ...`

File: cmd/thanos/receive.go [hunk 30]
```diff
@@ -0,0 +1,271 @@
...
+	level.Debug(logger).Log("msg", "setting up endpoint readiness")
+	{
+		// Initial configuration loading.
+		cancel := make(chan struct{})
+		g.Add(
+			func() error {
+				select {
+				case <-dbOpen:
+					break
+				// In case a shutdown is initiated before the dbOpen is released
```
Comment (bwplotka) [inline cmd/thanos/receive.go]: Missing trailing period in comment (:

Comment (bwplotka) [reply cmd/thanos/receive.go]: Also I would remove this, I think this is quite clear.

File: cmd/thanos/receive.go [hunk 31]
```diff
@@ -0,0 +1,271 @@
...
+
+	level.Debug(logger).Log("msg", "setting up metric http listen-group")
+	if err := metricHTTPListenGroup(g, logger, reg, httpMetricsBindAddr); err != nil {
+		return err
+	}
+
+	level.Debug(logger).Log("msg", "setting up grpc server")
+	{
+		var (
+			logger = log.With(logger, "component", "receiver")
```
Comment (bwplotka) [inline cmd/thanos/receive.go]: Why not doing this in very top?

File: go.sum [hunk 32]
```diff
@@ -204,6 +204,7 @@ github.com/onsi/ginkgo v1.6.0 h1:Ix8l273rp3QzYgXSR+c8d1fTG7UPgYkOSELPhiY/YGw=
 github.com/onsi/ginkgo v1.6.0/go.mod h1:lLunBs/Ym6LB5Z9jYTR76FiuTmxDTDusOGeTQH+WWjE=
 github.com/onsi/gomega v1.4.1 h1:PZSj/UFNaVp3KxrzHOcS7oyuWA7LoOY/77yCTEFu21U=
 github.com/onsi/gomega v1.4.1/go.mod h1:C1qb7wdrVGGVU+Z6iS04AVkA3Q65CEZX59MT0QO5uiA=
+github.com/opentracing-contrib/go-stdlib v0.0.0-20170113013457-1de4cc2120e7 h1:8KbikWulLUcMM96hBxjgoo6gTmCkG6HYSDohv/WygYU=
```
Comment (bwplotka) [inline go.sum]: modules :heart:

File: pkg/receive/handler.go [hunk 33]
```diff
@@ -0,0 +1,204 @@
...
+		logger:       logger,
+		router:       router,
+		readyStorage: o.ReadyStorage,
+		receiver:     o.Receiver,
+		options:      o,
+		quitCh:       make(chan struct{}),
+	}
+
+	readyf := h.testReady
+	router.Post("/receive", readyf(h.receive))
```
Comment (bwplotka) [inline pkg/receive/handler.go]: Ok, so this starts to be some API. We already keep Prometheus like API to guide people and be intuitive. This means that remote receiver would be nice to use the same or similar... does Prometheus serves remote write endpoint? What endpoint looks like?  `api/v1/write` ? As a basics something like `v1` in path is useful.

File: pkg/receive/handler.go (hunk 8 above)
Comment (bwplotka) [reply pkg/receive/handler.go]: Not addressed

File: pkg/receive/handler.go (hunk 33 above)
Comment (domgreen) [reply pkg/receive/handler.go]: tend to agree but flexible on the url ... `api/v1/receive` feels more correct to me

Comment (brancz) [reply pkg/receive/handler.go]: agreed with the prefix, going with `/api/v1/receive`

**Comment by brancz** [comment]
I believe I addressed everything:

* Added run group handling for the http handler
* Removed gossip
* Took the freedom to convert the quickstart script from gossip to static `--store` flag configuration

File: cmd/thanos/flags.go [hunk 34]
```diff
@@ -17,25 +17,37 @@ import (
 	"gopkg.in/alecthomas/kingpin.v2"
 )
 
-func regCommonServerFlags(cmd *kingpin.CmdClause) (
+func regGRPCFlags(cmd *kingpin.CmdClause) (
```
Comment (bwplotka) [inline cmd/thanos/flags.go]: hahah I was starring at monitor 1m and thinking why you removed Gossip for everyone in this PR... then seen that actually you added this function.... Github troll =D
